src/common/ad_mem.c

```diff
@@ -261,18 +292,84 @@ blob_fini(struct ad_blob *blob)
 	while ((arena = d_list_pop_entry(&blob->bb_ars_lru, struct ad_arena, ar_link)))
 		free_arena(arena, true);
 
+	d_binheap_destroy(&blob->bb_arena_free_heap);
+	D_FREE(blob->bb_mh_nodes);
 	D_FREE(blob->bb_bmap_rsv);
 	D_FREE(blob->bb_pages);
 	D_FREE(blob->bb_mmap);
 	return 0;
 }
 
+static uint32_t
+arena_free_size(struct ad_arena_df *ad)
+{
+	int		    grp_idx;
+	uint32_t	    weight = 0;
+	struct ad_group_df  *gd;
+	uint32_t	    grp_max = (ARENA_SIZE + ARENA_UNIT_SIZE - 1) / ARENA_UNIT_SIZE;
+
+	for (grp_idx = 0; grp_idx < grp_max; grp_idx++) {
+		if (!isset64(ad->ad_bmap, grp_idx)) {
+			weight += (1 << GRP_SIZE_SHIFT);
+			continue;
+		}
+	}
+
+	for (grp_idx = 0; grp_idx < ad->ad_grp_nr; grp_idx++) {
+		gd = &ad->ad_groups[grp_idx];
+		weight += gd->gd_unit_free * gd->gd_unit;
+	}
+
+	return weight;
+}
+
+#define ARENA_WEIGHT_BITS 12
+
+/* Avoid to change weight of an arena on every alloc/free (reorder arena) */
+static inline uint32_t
+arena_weight(int free_size)
+{
+	return (free_size + (1 << ARENA_WEIGHT_BITS) - 1) >> ARENA_WEIGHT_BITS;
+}
+
+static int
+arena_insert_free_entry(struct ad_blob *blob, uint32_t arena_id, int free_size)
+{
+	struct ad_maxheap_node	*mh_node;
+	int			 rc;
+
+	D_ASSERT(arena_id < blob->bb_pgs_nr);
+	mh_node = &blob->bb_mh_nodes[arena_id];
+
+	mh_node->mh_free_size = free_size;
+	mh_node->mh_weight = arena_weight(free_size);
+	mh_node->mh_arena_id = arena_id;
+	mh_node->mh_in_tree = 1;
```

Comment:
how about set mh_in_tree in hop_enter and hop_exit?
so we don't have to take care of it all the time.

Reply:
good idea